parts/linux/cloud-init/artifacts/cse_install.sh

```diff
@@ -200,105 +201,138 @@ downloadSecureTLSBootstrapKubeletExecPlugin() {
-    BOOTSTRAP_PROFILE_CONTAINER_REGISTRY_SERVER="${BOOTSTRAP_PROFILE_CONTAINER_REGISTRY_SERVER:=}"
+wasmFilesExist() {
+    local containerd_wasm_filepath=${1}
+    local shim_version=${2}
+    local version_suffix=${3}
+    shift 3
+    local shims_to_download=("$@")
+
+    local binary_version="$(echo "${shim_version}" | tr . -)"
+    for shim in "${shims_to_download[@]}"; do
+        if [ ! -f "${containerd_wasm_filepath}/containerd-shim-${shim}-${binary_version}-${version_suffix}" ]; then
+            return 1 # file is missing
+        fi
+    done
+    echo "all wasm files exist for ${containerd_wasm_filepath}/containerd-shim-*-${binary_version}-${version_suffix}"
+    return 0 
+}
+
+# Install, download, update wasm must all be run from the same function call
+# in order to ensure WASMSHIMPIDS persists correctly since in bash a new
+# function call from install-dependnecies will create a new shell process.
+installContainerdWasmShims(){
+    local download_location=${1}
+    PACKAGE_DOWNLOAD_URL=${2}
+    shift 2 # shift past the first 2 arguments to capture the list of versions
+    local package_versions=("$@")
     
-    for shim_version in $CONTAINERD_WASM_VERSIONS; do
-        binary_version="$(echo "${shim_version}" | tr . -)"
-
-        local version_suffix
-        local shims_to_download=()
-        local shim_prefix="containerd-shim-"
-        local registry_path
-        local base_path
-        local shim_filename
-
-        # figure out version suffix, shims to download, and paths
-        if [ "$shim_version" == "v0.15.1" ]; then
-            version_suffix="-v2"
-            shims_to_download=("spin")
-            registry_path="oss/binaries/spinkube/containerd-shim-spin"
-            base_path="spinkube"
-            shim_filename="containerd-shim-spin-v2"
-        else
-            version_suffix="-v1"
-            shims_to_download=("spin" "slight")
-            registry_path="oss/binaries/deislabs/containerd-wasm-shims"
-            base_path="containerd-wasm-shims"
-            shim_filename="containerd-wasm-shims-linux-${CPU_ARCH}.tar.gz"
-            if [ "$shim_version" == "v0.8.0" ]; then
-                shims_to_download+=("wws")
-            fi
+    for version in "${package_versions[@]}"; do
+        local shims_to_download=("spin" "slight")
+        if [[ "$version" == "0.8.0" ]]; then
+            shims_to_download+=("wws")
+        fi
+        containerd_wasm_url=$(evalPackageDownloadURL ${PACKAGE_DOWNLOAD_URL})
+        downloadContainerdWasmShims $download_location $containerd_wasm_url "v$version" "${shims_to_download[@]}" # adding v to version for simplicity
+    done
+    wait ${WASMSHIMPIDS[@]}
+    for version in "${package_versions[@]}"; do
+        local shims_to_download=("spin" "slight")
+        if [[ "$version" == "0.8.0" ]]; then
+            shims_to_download+=("wws")
         fi
+        updateContainerdWasmShimsPermissions $download_location "v$version" "${shims_to_download[@]}"
```

Comment:
Why not placing `updateContainerdWasmShimsPermissions` in the previous for-loop? Is it for optimizing the download speed?
If that's the case, could you please leave a comment to clarify? Thanks

Reply:
Yep exactly, we want to wait for the processes to finish downloading. I’ll leave a comment